### ISSUE TITLE
(P2)nxos_bgp_neighbor - several properties not idempotent

### ISSUE THREAD

**Issue by rahushen** [ansible/ansible#26262]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

- nxos_bgp_neighbor

##### ANSIBLE VERSION

```
 ansible 2.4.0 (devel 6bed831f8c) last updated 2017/06/27 15:24:19 (GMT -400)
  config file = None
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /root/agents-ci/ansible/lib/ansible
  executable location = /root/agents-ci/ansible/bin/ansible
  python version = 2.7.6 (default, Oct 26 2016, 20:30:19) [GCC 4.8.4]
```

##### OS / ENVIRONMENT

- Ansible Server : Ubuntu 14.04
- Device: Nexus9000 C9508  running 7.0(3)I6(1)

##### SUMMARY

Several properties are not idempotent

##### STEPS TO REPRODUCE

```yaml
---
cli:
  host: "{{ inventory_hostname }}"
  username: "{{ nxos_cli_user | default('admin') }}"
  password: "{{ nxos_cli_pass | default('admin') }}"

nxapi:
  host: "{{ inventory_hostname }}"
  username: "{{ nxos_nxapi_user | default('admin') }}"
  password: "{{ nxos_nxapi_pass | default('admin') }}"
  transport: nxapi
  user_ssl: no

---
- name: NXOS Sample Playbook
  hosts: nxos
  connection: local
  gather_facts: no

  tasks:

  - name: Ensure BGP Is Enabled
    nxos_feature:
      feature: bgp
      state: enabled
      provider: "{{ cli }}"

  - name: "Configure BGP neighbor"
    nxos_bgp_neighbor: &configure_non_default
      asn: 65535
      neighbor: 3.3.3.3
      description: "tested by ansible"
      connected_check: true
      capability_negotiation: true
      dynamic_capability: true
      ebgp_multihop: 2
      log_neighbor_changes: enable
      low_memory_exempt: true
      remote_as: 12.1
      remove_private_as: all
      suppress_4_byte_as: true
      update_source: loopback151
      state: present
      provider: "{{ cli }}"
```

##### EXPECTED RESULTS

expected playbook to be ok the second time instead of 'changed'.

##### ACTUAL RESULTS

```
PLAY [NXOS Sample Playbook] ***********************************************************************************************************************************************************************************************************************************

TASK [Ensure BGP Is Enabled] **********************************************************************************************************************************************************************************************************************************
ok: [nxxxx-xxx.example.com]

TASK [Configure BGP neighbor] *********************************************************************************************************************************************************************************************************************************
changed: [nxxxx-xxx.example.com]

PLAY RECAP ****************************************************************************************************************************************************************************************************************************************************
nxxxx-xxx.example.com          : ok=2    changed=1    unreachable=0    failed=0
```

**Comment by ansibot** [comment]
cc @jedelman8 @GGabriele @privateip @rcarrillocruz @mikewiebe @trishnaguha
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by s-hertel** [comment]
Hi @rahushen. What version of Ansible are you using?

**Comment by s-hertel** [comment]
needs_info

**Comment by rahushen** [comment]
@s-hertel updated issue with ansible version.

**Comment by Qalthos** [comment]
Could you please be more specific about which properties are not idempotent? (ie, by showing the contents of results.commands across successive runs)

**Comment by gundalow** [comment]
resolved_by_pr 27348

**Comment by rahushen** [comment]
@Qalthos ... here you go.

 TASK [nxos_bgp_neighbor : Check Idempotence] **************************************************************************************************************
 task path: /root/agents-ci/ansible/test/integration/targets/nxos_bgp_neighbor/tests/cli/sanity.yaml:81
 <n9k-xxx.example.com> connection transport is cli
 <n9k-xxx.example.com> using connection plugin network_cli
 <n9k-xxx.example.com> socket_path: /root/.ansible/pc/cb2b23dc81
 Using module file /root/agents-ci/ansible/lib/ansible/modules/network/nxos/nxos_bgp_neighbor.py
 <n9k-xxx.example.com> ESTABLISH LOCAL CONNECTION FOR USER: root
 <n9k-xxx.example.com> EXEC /bin/sh -c 'echo ~ && sleep 0'
 <n9k-xxx.example.com> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1501606780.44-175228690872683 `" && echo ansible-tmp-1501606780.44-175228690872683="` echo /root/.ansible/tmp/ansible-tmp-1501606780.44-175228690872683 `" ) && sleep 0'
 <n9k-xxx.example.com> PUT /tmp/tmp7cmY9u TO /root/.ansible/tmp/ansible-tmp-1501606780.44-175228690872683/nxos_bgp_neighbor.py
 <n9k-xxx.example.com> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1501606780.44-175228690872683/ /root/.ansible/tmp/ansible-tmp-1501606780.44-175228690872683/nxos_bgp_neighbor.py && sleep 0'
 <n9k-xxx.example.com> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1501606780.44-175228690872683/nxos_bgp_neighbor.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1501606780.44-175228690872683/" > /dev/null 2>&1 && sleep 0'
 changed: [n9k-xxx.example.com] => {
     "changed": true,
     "commands": [
         "router bgp 65535",
         "neighbor 3.3.3.3",
         "log-neighbor-changes",
         "remove-private-as all"
     ],
     "failed": false,
     "invocation": {
         "module_args": {
             "asn": "65535",
             "capability_negotiation": true,
             "connected_check": true,
             "description": "tested by ansible",
             "dynamic_capability": true,
             "ebgp_multihop": "2",
             "host": null,
             "local_as": null,
             "log_neighbor_changes": "enable",
             "low_memory_exempt": true,
             "m_facts": false,
             "maximum_peers": null,
             "neighbor": "3.3.3.3",
             "password": null,
             "port": null,
             "provider": null,
             "pwd": null,
             "pwd_type": null,
             "remote_as": "12.1",
             "remove_private_as": "all",
             "shutdown": true,
             "ssh_keyfile": null,
             "state": "present",
             "suppress_4_byte_as": true,
             "timeout": 10,
             "timers_holdtime": "270",
             "timers_keepalive": "90",
             "transport": "cli",
             "transport_passive_only": null,
             "update_source": "loopback151",
             "use_ssl": null,
             "username": null,
             "validate_certs": null,
             "vrf": "default"
         }
     }
 }

 TASK [nxos_bgp_neighbor : assert] *************************************************************************************************************************
 task path: /root/agents-ci/ansible/test/integration/targets/nxos_bgp_neighbor/tests/cli/sanity.yaml:85
 FAILED! => {
     "assertion": "result.changed == false",
     "changed": false,
     "evaluated_to": false,
     "failed": true
 }

**Comment by Qalthos** [comment]
Well that makes sense, the ones I can't test aren't idempotent. It looks like we have a version compatibility issue here

**Comment by rahushen** [comment]
@Qalthos ... this is the run on N7K hardware 

TASK [nxos_bgp_neighbor : Configure BGP neighbor non-defaults] ********************************************************************************************
task path: /root/agents-ci/ansible/test/integration/targets/nxos_bgp_neighbor/tests/cli/sanity.yaml:57
<n7k-xxx.example.com> connection transport is cli
<n7k-xxx.example.com> using connection plugin network_cli
<n7k-xxx.example.com> socket_path: /root/.ansible/pc/580e811d76
Using module file /root/agents-ci/ansible/lib/ansible/modules/network/nxos/nxos_bgp_neighbor.py
<n7k-xxx.example.com> ESTABLISH LOCAL CONNECTION FOR USER: root
<n7k-xxx.example.com> EXEC /bin/sh -c 'echo ~ && sleep 0'
<n7k-xxx.example.com> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1501699566.75-135754821422360 `" && echo ansible-tmp-1501699566.75-135754821422360="` echo /root/.ansible/tmp/ansible-tmp-1501699566.75-135754821422360 `" ) && sleep 0'
<n7k-xxx.example.com> PUT /tmp/tmpnYUCre TO /root/.ansible/tmp/ansible-tmp-1501699566.75-135754821422360/nxos_bgp_neighbor.py
<n7k-xxx.example.com> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1501699566.75-135754821422360/ /root/.ansible/tmp/ansible-tmp-1501699566.75-135754821422360/nxos_bgp_neighbor.py && sleep 0'
<n7k-xxx.example.com> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1501699566.75-135754821422360/nxos_bgp_neighbor.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1501699566.75-135754821422360/" > /dev/null 2>&1 && sleep 0'
changed: [n7k-xxx.example.com] => {
    "changed": true,
    "commands": [
        "router bgp 65535",
        "neighbor 3.3.3.3",
        "dynamic-capability",
        "low-memory exempt",
        "update-source loopback151",
        "log-neighbor-changes",
        "dont-capability-negotiate",
        "timers 90 270",
        "remove-private-as all",
        "shutdown",
        "disable-connected-check",
        "ebgp-multihop 2",
        "remote-as 12.1",
        "capability suppress 4-byte-as",
        "description tested by ansible"
    ],
    "failed": false,
    "invocation": {
        "module_args": {
            "asn": "65535",
            "capability_negotiation": true,
            "connected_check": true,
            "description": "tested by ansible",
            "dynamic_capability": true,
            "ebgp_multihop": "2",
            "host": null,
            "local_as": null,
            "log_neighbor_changes": "enable",
            "low_memory_exempt": true,
            "m_facts": false,
            "maximum_peers": null,
            "neighbor": "3.3.3.3",
            "password": null,
            "port": null,
            "provider": null,
            "pwd": null,
            "pwd_type": null,
            "remote_as": "12.1",
            "remove_private_as": "all",
            "shutdown": true,
            "ssh_keyfile": null,
            "state": "present",
            "suppress_4_byte_as": true,
            "timeout": null,
            "timers_holdtime": "270",
            "timers_keepalive": "90",
            "transport": "cli",
            "transport_passive_only": null,
            "update_source": "loopback151",
            "use_ssl": null,
            "username": null,
            "validate_certs": null,
            "vrf": "default"
        }
    }
}

TASK [nxos_bgp_neighbor : assert] *************************************************************************************************************************
task path: /root/agents-ci/ansible/test/integration/targets/nxos_bgp_neighbor/tests/cli/sanity.yaml:79
ok: [n7k-xxx.example.com] => {
    "changed": false,
    "failed": false,
    "msg": "All assertions passed"
}

TASK [nxos_bgp_neighbor : Check Idempotence] **************************************************************************************************************
task path: /root/agents-ci/ansible/test/integration/targets/nxos_bgp_neighbor/tests/cli/sanity.yaml:81
<n7k-xxx.example.com> connection transport is cli
<n7k-xxx.example.com> using connection plugin network_cli
<n7k-xxx.example.com> socket_path: /root/.ansible/pc/580e811d76
Using module file /root/agents-ci/ansible/lib/ansible/modules/network/nxos/nxos_bgp_neighbor.py
<n7k-xxx.example.com> ESTABLISH LOCAL CONNECTION FOR USER: root
<n7k-xxx.example.com> EXEC /bin/sh -c 'echo ~ && sleep 0'
<n7k-xxx.example.com> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1501699572.98-67475249209184 `" && echo ansible-tmp-1501699572.98-67475249209184="` echo /root/.ansible/tmp/ansible-tmp-1501699572.98-67475249209184 `" ) && sleep 0'
<n7k-xxx.example.com> PUT /tmp/tmpmIjZYw TO /root/.ansible/tmp/ansible-tmp-1501699572.98-67475249209184/nxos_bgp_neighbor.py
<n7k-xxx.example.com> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1501699572.98-67475249209184/ /root/.ansible/tmp/ansible-tmp-1501699572.98-67475249209184/nxos_bgp_neighbor.py && sleep 0'
<n7k-xxx.example.com> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1501699572.98-67475249209184/nxos_bgp_neighbor.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1501699572.98-67475249209184/" > /dev/null 2>&1 && sleep 0'
changed: [n7k-xxx.example.com] => {
    "changed": true,
    "commands": [
        "router bgp 65535",
        "neighbor 3.3.3.3",
        "log-neighbor-changes",
        "remove-private-as all"
    ],
    "failed": false,
    "invocation": {
        "module_args": {
            "asn": "65535",
            "capability_negotiation": true,
            "connected_check": true,
            "description": "tested by ansible",
            "dynamic_capability": true,
            "ebgp_multihop": "2",
            "host": null,
            "local_as": null,
            "log_neighbor_changes": "enable",
            "low_memory_exempt": true,
            "m_facts": false,
            "maximum_peers": null,
            "neighbor": "3.3.3.3",
            "password": null,
            "port": null,
            "provider": null,
            "pwd": null,
            "pwd_type": null,
            "remote_as": "12.1",
            "remove_private_as": "all",
            "shutdown": true,
            "ssh_keyfile": null,
            "state": "present",
            "suppress_4_byte_as": true,
            "timeout": null,
            "timers_holdtime": "270",
            "timers_keepalive": "90",
            "transport": "cli",
            "transport_passive_only": null,
            "update_source": "loopback151",
            "use_ssl": null,
            "username": null,
            "validate_certs": null,
            "vrf": "default"
        }
    }
}

TASK [nxos_bgp_neighbor : assert] *************************************************************************************************************************
task path: /root/agents-ci/ansible/test/integration/targets/nxos_bgp_neighbor/tests/cli/sanity.yaml:85
FAILED! => {
    "assertion": "result.changed == false",
    "changed": false,
    "evaluated_to": false,
    "failed": true
}

**Comment by rahushen** [comment]
@Qalthos .... manually applying the configs to the N7K hardware

n7k-99(config)# feature bgp
n7k-99(config)# router bgp 65535
n7k-99(config-router)# neighbor 3.3.3.3
n7k-99(config-router-neighbor)# dynamic-capability
n7k-99(config-router-neighbor)# low-memory exempt
n7k-99(config-router-neighbor)# update-source loopback151
n7k-99(config-router-neighbor)# log-neighbor-changes
n7k-99(config-router-neighbor)# dont-capability-negotiate
n7k-99(config-router-neighbor)# timers 90 270
n7k-99(config-router-neighbor)# remove-private-as all
n7k-99(config-router-neighbor)# shutdown
n7k-99(config-router-neighbor)# disable-connected-check
n7k-99(config-router-neighbor)# ebgp-multihop 2
n7k-99(config-router-neighbor)# remote-as 12.1
n7k-99(config-router-neighbor)# capability suppress 4-byte-as
n7k-99(config-router-neighbor)# description tested by ansible
n7k-99(config-router-neighbor)#

**Comment by rahushen** [comment]
@Qalthos .... sh running bgp output after applying configs

n7k-99# sh running bgp

!Command: show running-config bgp
!Time: Wed Aug  2 14:53:39 2017

version 8.0(1)
feature bgp

router bgp 65535
  neighbor 3.3.3.3
    low-memory exempt
    remote-as 12.1
    log-neighbor-changes
    description tested by ansible
    shutdown
    dont-capability-negotiate
    update-source loopback151
    remove-private-as all
    disable-connected-check
    capability suppress 4-byte-as
    ebgp-multihop 2
    timers 90 270

**Comment by Qalthos** [comment]
Okay. I fixed `log-neighbor-changes`. I also changed nxos_bgp_neighbor so that warnings printed on the cli are reported back to the user.

This playbook is still not idempotent for me, but at least now there is a warning being returned to the user in this case. I _believe_ the module should detect the config properly now, should it be present, but I do not know how to make it stick in our environment. I have, however, added a unit test to check that the right logic is being applied. Hopefully the real thing works as well.

Let me know if this works for you
